synapse/handlers/e2e_keys.py

```diff
@@ -976,16 +984,128 @@ def _get_e2e_cross_signing_verify_key(self, user_id, key_type, from_user_id=None
 
         Raises:
             NotFoundError: if the key is not found
+            SynapseError: if `user_id` is invalid
         """
+        user = UserID.from_string(user_id)
+        key_id = None
+        verify_key = None
+
         key = yield self.store.get_e2e_cross_signing_key(
             user_id, key_type, from_user_id
         )
+
+        # If we couldn't find the key locally, and we're looking for keys of
+        # another user then attempt to fetch the missing key from the remote
+        # user's server.
+        #
+        # We may run into this in possible edge cases where a user tries to
+        # cross-sign a remote user, but does not share any rooms with them yet.
+        # Thus, we would not have their key list yet. We fetch the key here,
+        # store it and notify clients of new, associated device IDs.
+        if (
+            key is None
+            and not self.is_mine(user)
+            # We only get "master" and "self_signing" keys from remote servers
+            and key_type in ["master", "self_signing"]
+        ):
+            key = yield self._retrieve_cross_signing_keys_for_remote_user(
```

Comment:
`_retrieve_cross_signing_keys_for_remote_user` (sometimes) returns a 3-tuple: are you handling that correctly?